### ISSUE TITLE
Executing workflows as a reviewer or reader

### ISSUE THREAD

**Issue by nuest** [alan-turing-institute/the-turing-way#1249]
## Detailed description

* This issue is a place to discuss matters relating to the writing of the chapter on _Executing workflows as a reviewer or reader_.
* Academics are often faced with challenges when they need to reproduce a workflow. This can happen as part of a peer review, or just to better understand someone else's work. Examples:
  * Reproducing a workflow can be highly educational and is one of the motivations behind **ReproHack** (https://n8cir.org.uk/news/reprohacks/).
  * The importance of reproducing computations gains more and more attention in peer review, but is still to rare (besides reproduction specific journals, e.g. ReScience C, and some lighthouse conference or journals). This is one of the motivations behind CODECHECK (https://codecheck.org.uk/).
  * (There's actually quite a few conferences and journals - I suggest to reach out to the reproducibility chairs or editors with a first draft of this chapter.)
  * While there is some common understanding on how to approach a peer review (though too often not written down), there is no common practice yet for approaching a reproducible workflow (shameless plug: [How to read a research compendium](https://arxiv.org/abs/1806.09525)).
* These initiatives and activities have specific needs, but also share a common and hard to obtain skill set: **How to quickly grasp the layout of a computational project, install the required software, and execute the actual workflow?**
* Authors of workflows are often experienced users in a particular package manager or workflow system, and it is hard to document to a suitable level for someone else (e.g., on a different operating system). Ideally, every paper's code would have a README that anyone can follow to reproduce the workflow, but it's hard to accomplish that (on a researcher's time budget). In addition, common problems (dependency hell, differences between operating systems, ..) make this hard in practice.
* I suggest to use the term "executing" instead of "reproducing", to make this more about the technical steps. _As a person executing the code, I don't need to understand what it does, or if it works._ But I can check if it runs without errors and if it creates outputs that I can find in the paper. _Then_ (second step), I know that things work and I can a) be happy with that and move on, or b) know that it is worth digging deeper and trying to understand things.
- This chapter could indirectly provide guidance for authors "how far" they should document.

### Table of contents

1. Computational workflow basics (basic terms, so that we can translate these terms to the different tools below)
1. Dependency management for common languages (if you see a file "X" then try out these commands)
  - Python (`requirements.txt`, `environment.yml`)
  - R (`DESCRIPTION`, `renv.lock`)
  - Julia (..)
  - Octave (..)
  - Node.js (`package.json`)
  - `Makefile`
1. Workflow/pipeline tools (same as above)
1. Containerisation (concrete images that can get you running quickly; how not to break your own system)
1. When to stop? (when is it more reasonable to reach out to the author)
1. How to give feedback (depending on the status of the paper, etc.)
1. Example reproduction reports (what did others execute, what problems did they have, how did they solve them; sources: ReproHack follow-ups, AGILE reproduction reports, CODECHECK reports)

### Resources

- https://www.software.ac.uk/blog/2019-08-07-reprohacking-carpentryconnect-manchester-2019
- https://codecheck.org.uk/guide/community-process
- Section "Reproducibility Reviewer" in [AGILE Reproducible Paper guidelines](https://docs.google.com/document/d/1Kc-ToUVcrdsq6aB8Qy2J_rIluFwDniv6GHGtZuPvIEo/edit#heading=h.8xtjt4j1hjwp)

### Discussion points/related issues

#1240 is related, though has a broader scope.

I've previously discussed the idea of collaborating on documentation for "people who reproduce papers" in the context of ReproHack (:wave: @annakrystalli) and CODECHECK (:wave: @sje30), and recently also with @HeidiSeibold. I've reproduced a few papers for a conference recently, and want to help the next generation of reproducibility reviewers, though part of the content would not be specific to that conference/discipline. We've been thinking about "train the trainer"-style of events, and would also need some documentation for that.

So I thought "maybe The Turing Way :heart: is a good venue". _What do you think?_

I see the following **challenges**

- Create something useful, without being overly specific or having to update weekly because of changes in tools/platforms.
- Not replicate documentation from package managers, but also not just send readers to long and complex documentation websites that might overwhelm them.
- Not sure how much can be documented (without being outdated quickly), and how much is (and will have to be) just experience.
- I said _"Have a README - all else is details."_ in a short talk I gave at the Remote ReproHack, so _maybe this chapter is actually a bad idea._

## Current status

This is an idea for a new chapter. @nuest looks for feedback and collaborators. If anyone would like to make a start they are more than welcome to do so.

 - [ ] Write chapter outline
 - [ ] Add material to the chapter
 - [ ] Combine materials into a readable chapter
 - [ ] Proofread
 - [ ] Request reviews
 - [ ] Address reviews
 - [ ] Merge to master branch.

## Updates

- 2020-07-03: Posted issue with first idea, tagging people who I have previously interacted with on this topic

**Comment by sje30** [comment]
Count me in @nuest !

**Comment by kkmann** [comment]
I feel it is a great idea to view the problem of reproducibility from different perspectives. So far, TTW is mainly written from an analysts perspective.

There is a certain risk of duplicating work here though. Maybe it could be worthwhile to rely on heavy cross linking to the respective existing sections - many points are already addressed... or at least 'outlined' ;)

Related but slightly different thought: it would be great to come up with more examples and how readme files should be structured independently of the software used to make it easier to get started.

**Comment by annakrystalli** [comment]
I agree with @kkmann thay alot of the proposed content is contained elsewhere (although pulling it all together at a high level in the research compendia chapter could be cool).

However @malvikasharan had previously mentioned a chapter on reviewing research compendia to go along with the research compendia chapter.

It feels this is were this particular group could add value. A number of the Reprohack team would be interested in contributing to something like this also.

**Comment by HeidiSeibold** [comment]
Exciting! I'd be happy to help :nerd_face: :sparkles: (after the summer though)

**Comment by nuest** [comment]
Thanks for the feedback @kkmann @annakrystalli and the support @sje30 @HeidiSeibold ! (after the summer sounds good)

I agree there is overlap with other sections, and I'd be happy to flesh out a little bit what is missing for this particular perspective, or make this chapter a well curated list of links to other chapters.

However, _I suggest to clearly distinguish execution from review/evaluation_. In this chapter, I imagined more practical advise to go through the _mechanics_ to run a particular workflow. "If you see requirements.txt, you need a virtual Python environment and then run pip install -e". For an execution, I do not need to understand what is happening, which is the case for a review of software/workflow.

@kkmann: On structuring READMEs: sure, could be helpful. But nothing will be as helpful as a second person actually trying to follow the instructions. Unless you're extremely careful/diligent, you will always forget to mention one thing that is so clear to you that you cannot imagine someone else not knowing. _So, yes, related, but maybe a different_ perspective _for another chapter?_

**Comment by kkmann** [comment]
ha, so true. the quality improvement in docs and code when going from one coder to two is huge. same goes for readme files - having a 'peer readyours' should be item one on the list of how to write good readme files ;)

**Comment by nuest** [comment]
@malvikasharan Can you share a bit more about the chapter on reviewing research compendia. I might go in a very similar direction than this idea.

**Comment by malvikasharan** [comment]
Hey @nuest, I think we discussed that in some issues around creating case studies that use research compendia and reuse that in varifying reproducibility, for example by using Anna's  ReproHack model. You should go ahead and start developing and documenting your idea and hopefully, we can invite some other folks for reviewing.

**Comment by nuest** [comment]
I hope to get to this when it get's a bit quieter towards the end of the year :-) :christmas_tree:
